Startx improvements

Atemu opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions
-->

###### Motivation for this change
Typo-less and improved version of #80272

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

Atemu committed:
xinit: use the system xserverrc

Previously it'd try to use the one under its output path which is read-only of
course

Atemu committed:
nixos/startx: provide xserverArgs via xserverrc

Fixes #80198

Atemu committed:
nixos/xserver: make logFile configurable

It makes sense for it to be /dev/null for all the displayManagers but startx,
it needs a different logFile configuration.

Atemu committed:
nixos/startx: send Xorg log to the default location

This partially reverts bf3d3dd19b48c432dd83aa0385b47dbe84aa647b.

I don't know why we weren't getting a default logfile back then but Xorg
definitely provides one now ($XDG_DATA_HOME for regular users and /var/log for
root, see `man Xorg`)

Atemu committed:
xorgserver: set log-dir to /var/log instead of the Nix store

Xorg creates the log-dir in its output path because X crashes if it can't write
to its logfile. On a regular distro, this dir would be installed into the root
to prevent that from happening but with Nix, it sits in the read-only Nix store.
Ironically, when Xorg tries to write here, it fails and crashes.

To make Xorg log to /var/log, we have to stop the build script from trying to
create the log-dir as the sandbox doesn't (and shouldn't) have access to /var.

This creates a runtime dependency on /var when running as root but that should
exist on any Linux system (on NixOS, journald always creates /var/log).

Previously, the startx displayManager required some workarounds for logfiles
which are obsolete now.

patchPhase -> postPatch because overriding the patchPhase prevents patches from
being applied

Atemu commented:
/marvin opt-in
/status needs_reviewer

doronbehar commented on nixos/modules/services/x11/xserver.nix:
```diff
@@ -523,6 +523,16 @@ in
+          Controls the file Xorg logs to.
+        '';
+      };
+
```
There should be a tmpfiles rules along in this commit.

doronbehar replied:
Then I guess that not giving a tmpfiles rules is a sort of discouragement to set this option to something else.

doronbehar reviewed (changes requested):
Diff looks good.

doronbehar commented:
The first 2 commits are totally fine I suppose. Regarding the log file's location changing from `~/.xorg.log` to `/var/log/xorg.log`, are we sure that won't cause any issues?

Atemu commented:
This changes it to the default upstream Xorg logfile location. The logfile was hard-coded by us to a non-standard location because of an issue someone had in 2012 which was likely caused by Nixpkgs patching the log location to be inside the Nix store, which I also fixed.

The upstream default log location for regular users is inside their home dir, /var/log will only be used when manually running X as root.

I really don't think we need a tmpfiles rule for this. Xorg logs are kilobytes and are rotated automatically AFAICT. They're aren't even going to be there in most cases as you need to *manually* run X as root.     

You probably don't even want them to be deleted automatically as you might want to audit them after a catastrophic failure.

If you `sudo startx` for some strange reason and care about a few kilobytes in /var/log, you should just delete them manually.

doronbehar reviewed (approved):
I approve as all changes make sense. But I didn't test so I don't merge.

timokau commented:
Sorry for the [spam](https://github.com/timokau/marvin-mk2/issues/99). Unfortunately I can't look into this in much detail myself right now. @Atemu to what extent have you tested this?

Atemu commented:
I sucessfully ran i3 with startx and the NixOS test still works too. I also tested running X as root.

Since this requires a rebuild of all Xorg dependent software, I didn't test much further but I could do that now.

Is there a specific scenario you would like to see tested?

timokau commented on nixos/modules/services/x11/xserver.nix:
```diff
@@ -523,6 +523,16 @@ in
 
+      logFile = mkOption {
+        type = types.nullOr types.str;
+        # Log at the default verbosity level to stderr rather than /var/log/X.*.log.
```
It would be nice to have this in the description instead of a comment ("pass /dev/null to log to stderr").

timokau commented on nixos/modules/services/x11/display-managers/startx.nix:
```diff
@@ -39,6 +39,14 @@ in
     systemd.services.display-manager.enable = false;
+
+    # Let Xorg use its default log location ($XDG_DATA_HOME/xorg/)
+    services.xserver.logFile = mkDefault null;
```
Why do we only do this when using a display manager?

timokau replied:
I found this commit which originally set the `/dev/null` default: d84741a4bfb6. Is there a reason to change the default here?

Atemu replied:
The Startx "displayManager" is not a systemd service and therefore doesn't log to journal when the logfile is set to /dev/null IIRC.

Startx logging to your home dir is what users would expect it to do.

timokau replied:
That makes a lot of sense :) Please add that reasoning in a comment as well.

timokau reviewed (commented):
That's probably sufficient, thank you :) I left some additional feedback.

Atemu commented on nixos/modules/services/x11/xserver.nix:
```diff
@@ -523,6 +523,16 @@ in
+        default = "/dev/null";
+        example = "/var/log/Xorg.0.log";
+        description = ''
+          Controls the file Xorg logs to.
```
Might as well document these behaviours properly, good point.

Atemu commented on pkgs/servers/x11/xorg/overrides.nix:
```diff
@@ -638,6 +638,7 @@ self: super:
       if (!isDarwin)
       then {
         outputs = [ "out" "dev" ];
+        patches = [ ./dont-create-logdir.patch ];
```
The commit message has a rather lengthy explanation already, I didn't want to bloat the code.

`dont-create-logdir-during-build` would be a better name for the patch though.

timokau replied:
Yes, the commit message is very helpful. But its even nicer to have the "why" at a glance, without having to dig into it. It doesn't have to go into any detail, just `# the default logdir is outside the nix store and will not be accessible`.

Ericson2314 replied:
@Atemu This should be explained in the code, not the commit message:

1. Understanding the new state of the code in isolation should be in code comments
2. Understanding the new state in relation to the old state should be in the commit message.

This cases falls into case 1, because someone can wonder what the patch is for at any point. Only answers to questions like "I understand the new and old way of doing things, but why is the new way better?" go in commit messages.

timokau commented on pkgs/servers/x11/xorg/overrides.nix:
```diff
@@ -654,6 +660,7 @@ self: super:
           "--with-xkb-bin-directory=${self.xkbcomp}/bin"
           "--with-xkb-path=${self.xkeyboardconfig}/share/X11/xkb"
           "--with-xkb-output=$out/share/X11/xkb/compiled"
+          "--with-log-dir=/var/log"
```
Why do we override this at compile-time? It's probably not necessary to change anything here (except maybe add a clarification in a comment, depending on the reason), just asking out of curiosity.

Atemu replied:
It defaults to $out/var and that seems to get hard-coded into the binary IIRC. Not a good location for trying to write logfiles, especially when a write failure causes a crash ;)

timokau replied:
Okay. Its always [preferable](https://mtlynch.io/code-review-love/#4-answer-questions-with-the-code-itself) to explain those things in a comment, but it's clear enough here and not important enough for another round of review. Thank you for the explanation :)

timokau reviewed (changes requested):
Looks good now, please squash your fixups.

Slightly off-topic: Its usually easier if you explicitly re-request a review (or post a comment) when you have addressed all comments. GitHub doesn't send notifications when threads are resolved.

Atemu commented:
Squashed and rebased. Also changed base to staging since this requires a rebuild of all Xorg-server dependent software for something most people don't use.

Atemu commented:
EditorConfig fail is unrelated, probably an issue in staging.

timokau reviewed (approved):
Thank you!

Atemu commented:
Thank you for the thorough review!
